presto-client/src/client.ts

```diff
@@ -0,0 +1,96 @@
+import { PrestoClientConfig, PrestoQuery, PrestoResponse } from './client.types'
+
+export class PrestoClient {
+  private baseUrl: string
+  private catalog: string
+  private headers: Record<string, string>
+  private interval: number
+  private schema: string
+  private timezone?: string
+  private user: string
+
+  constructor({ catalog, host, interval, port, schema, timezone, user }: PrestoClientConfig) {
+    this.baseUrl = `${host || 'http://localhost'}:${port || 8080}/v1/statement`
+    this.catalog = catalog
+    this.interval = interval || 100
+    this.schema = schema
+    this.timezone = timezone
+    this.user = user
+
+    this.headers = { 'X-Presto-Client-Info': 'presto-js-client' }
+    if (this.user) {
+      this.headers['X-Presto-User'] = this.user
+    }
+    if (this.schema) {
+      this.headers['X-Presto-Schema'] = this.schema
+    }
+    if (this.catalog) {
+      this.headers['X-Presto-Catalog'] = this.catalog
+    }
+    if (this.timezone) {
+      this.headers['X-Presto-Time-Zone'] = this.timezone
+    }
+
+    // TODO: Set up auth
+  }
+
+  private request(url: string, method: string, body: string | null = null) {
+    return fetch(url, {
+      body,
+      headers: this.headers,
+      method,
+    })
+  }
+
+  async query(query: string): Promise<PrestoQuery> {
+    // TODO: Checks for valid schema and catalog
+    const firstResponse = await this.request(this.baseUrl, 'POST', query)
+
+    if (firstResponse.status !== 200) {
+      throw new Error(`Query failed: ${JSON.stringify(await firstResponse.text())}`)
+    }
+
+    let nextUri = ((await firstResponse?.json()) as PrestoResponse)?.nextUri
+
+    if (!nextUri) {
+      throw new Error(`Didn't receive the first nextUri`)
+    }
+
+    let queryId: string | undefined = undefined
+    const columns = []
+    const data = []
+
+    do {
+      const response = await this.request(nextUri, 'GET')
+
+      if (response.status !== 200) {
+        throw new Error(`Query failed: ${JSON.stringify(await response.text())}`)
+      }
+
+      const prestoResponse = (await response.json()) as PrestoResponse
+
+      nextUri = prestoResponse?.nextUri
```

Comment:
need to verify if the last response's nextUri also point to itself but not undefined/null. I assume you already verify it. just want to double check.

Reply:
the undefined is checked in the while condition on line 84
I don't think we're checking if the nextUri is the same. Is that possible in Presto? if so I'll add that check too

Reply:
The nextUri will just return empty as the client has already consumed results if the URI is precisely the same.  The nextUri ought to be auto-incrementing though.  IOW safe to not check that.

Reply:
thanks @tdcmeehan for chiming in. I was thinking about tagging you while I had that comment. I double-checked that this morning. The last response doesn't have `nextUri`.